### PR TITLE
[Patch-t5-tokenizer] Patches the changes on T5 to make sure previous behaviour is still valide for beginning of words

### LINKED ISSUE #24569: LlamaTokenizer: Slow implementation opts for whitespace-lead token (different from fast)
### System Info

- `transformers` version: 4.30.2
- Platform: Linux-5.15.0-75-generic-x86_64-with-glibc2.31
- Python version: 3.10.11
- Huggingface_hub version: 0.14.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker @youn

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Reproduction

Comparing slow and fast `LlamaTokenizer` instances with `huggyllama/llama-7b`.

```
from transformers import AutoTokenizer

model = "huggyllama/llama-7b"

fast = AutoTokenizer.from_pretrained(model)
slow = AutoTokenizer.from_pretrained(model, use_fast=False)

# use tokenize()
print(fast.tokenize("<s>uns"), slow.tokenize("<s>uns"))
# -> (['▁<s>', 'uns'], ['<s>', '▁uns'])

# use __call__
print(fast(f"{fast.bos_token}uns", add_special_tokens=False), slow(f"{slow.bos_token}uns", add_special_tokens=False))
# -> ({'input_ids': [1, 6948], 'token_type_ids': [0, 0], 'attention_mask': [1, 1]},
#     {'input_ids': [1, 9644], 'attention_mask': [1, 1]})

# round-tripping
print(fast.convert_tokens_to_string(fast.tokenize("<s>uns")), fast.convert_tokens_to_string(slow.tokenize("<s>uns")))
# -> ('<s>uns', '<s> uns')
```

### Expected behavior

It looks like the slow LlamaTokenizer wrongly tokenises `uns`. I would not expect the additional whitespace when round-tripping or when tokenising in the first place.

Thanks a lot in advance.

### LINKED ISSUE #24743: T5 Tokenizer Adds Space after Each Added (Extra) Token
### System Info

- `transformers` version: 4.30.2
- Platform: Linux-5.4.0-146-generic-x86_64-with-glibc2.35
- Python version: 3.11.3
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.1+cu117 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: (NA)
- Using distributed or parallel set-up in script?: (NA)

### Who can help?

@Arthu

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```ipython
In [1]: from transformers import AutoTokenizer

In [2]: tokenizer = AutoTokenizer.from_pretrained("./models/t5-base/")

In [3]: tokenizer.add_tokens(["asdfg"], special_tokens=False)
Out[3]: 1

In [4]: tokenizer.tokenize("asdfgwordtimeasdfgtime")
Out[4]: ['asdfg', '▁word', 'time', 'asdfg', '▁time']
```

### Expected behavior

tokenizer return `['asdfg', 'word', 'time', 'asdfg', 'time']`

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
There was a small typo that modified the behaviour in #24565, the test were not able to catch it.  #24569
When a sentence does not start with a space, a space was added.
Before:
```python
>>>tokenizer.tokenize("Hello <extra_id_0>")
['_', '_Hello', '<extra_id_0>']
```
After:
```python 
>>>tokenizer.tokenize("Hello <extra_id_0>")
['_Hello', '<extra_id_0>']
```
# Big bug 35 models involved
Not only punctuation but anything after a special token is basically wrong... Let's ignore the fact that we also split when it's the beginning of word less important
image

Tests were added as they were green before merging

### CODE REVIEW

File: src/transformers/models/t5/tokenization_t5.py
Comment (ArthurZucker) [inline src/transformers/models/t5/tokenization_t5.py:305]: we always prepend with `SPIECE_UNDERLINE`. That way we can know id we have to post process the string or not. 
This is in order to `hack` the `_tokenize` method, which does not know if we are at the beginning of the string or if we are after a special token, which eats anything

**Comment by ArthurZucker** [comment]
Ran all the tests with `RUN_SLOW`, switch-ci is fixed

File: src/transformers/models/t5/tokenization_t5.py
Comment (amyeroberts) [inline src/transformers/models/t5/tokenization_t5.py:312]: We don't need to check `text.startswith(SPIECE_UNDERLINE)` again, as there's an early return on L310, so we know all text here doesn't start with `SPIECE_UNDERLINE`

```suggestion
        if tokens[0] == SPIECE_UNDERLINE:
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! Fix looks OK to me. I'm not super familiar with tokenizers, so would be good to have a second approval before merging. 

Do we have equivalent tests for other tokenizers in the library?

File: src/transformers/models/t5/tokenization_t5.py
Comment (ArthurZucker) [reply src/transformers/models/t5/tokenization_t5.py]: Indeed!

**Review by Narsil — COMMENTED** [review]
I don't feel good about this kind of change.

Adding `ifs` in general in tokenization methods is asking for trouble as the number of differents cases is astronomical (we have too much surface).

I wasn't a bit fan on the original fix either.

IMO we need to strengthen way more the test suite.

File: src/transformers/models/t5/tokenization_t5.py
Comment (Narsil) [inline src/transformers/models/t5/tokenization_t5.py:304]: No ifs, this is good !

File: src/transformers/models/t5/tokenization_t5.py
Comment (Narsil) [inline src/transformers/models/t5/tokenization_t5.py:310]: Why don't we want the follow-up logic ?

Comment (ArthurZucker) [reply src/transformers/models/t5/tokenization_t5.py:310]: will add lots of comments about this!

**Comment by ArthurZucker** [comment]
Until the added tokens are fixed, this will break the slow version that use extra ids, because by default we strip left and right..... So buggy

File: tests/models/t5/test_tokenization_t5.py
Comment (Narsil) [inline tests/models/t5/test_tokenization_t5.py:469]: Can we use the same string ?

File: tests/models/t5/test_tokenization_t5.py
Comment (Narsil) [inline tests/models/t5/test_tokenization_t5.py:489]: Test it, instead of putting it in comments.

Why not using the same string ? I'm lost as to the interest of testing 2 different strings.
We're testing the last 2 tokens are the same, but should ALL tokens be the same ?

If not, assert both ids, and explain in a comment why the values are not the same instead.

File: tests/models/t5/test_tokenization_t5.py
Comment (Narsil) [inline tests/models/t5/test_tokenization_t5.py:507]: Why aren't we testing against XNLI ?

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [reply tests/models/llama/test_tokenization_llama.py:561]: I'll add a comment, the idea is to make sure that the result of `special` next to `I` will give the same id of `I` than if `I` was next to a known token. Just making sure we don't add an extra space, but anyway tokenize suffices will remove

**Review by Narsil — APPROVED** [review]
LGTM.

- `legacy` means we are not modifying existing tokenizers without knowing. (And we need to manually update those core tokenizers)
- Test suite is more extensive and if we make code modifications later I feel more confident we're note going to rebreak things unknowingly (either way).

File: src/transformers/models/llama/tokenization_llama.py
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:139]: nit: formatting / spelling 
```suggestion
        Returns a tokenized string. 
        
        Since the sentencepiece internal model always adds a SPIECE_UNDERLINE, at the
        beginning of the provided text, we need to remove it by hand when the current text is a subsequence. This
        happens whenever the `self.tokenize` function is called with specials tokens: the input is split on the special
        tokens, and each subsequence is passed to `_tokenize`. Thus if a subsequence did not start with a `" "` or
        SPIECE_UNDERLINE, we have to remove the extra `SPIECE_UNDERLINE` prepended.
```

File: src/transformers/models/llama/tokenization_llama.py
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:149]: I don't believe you need the check here, I can do a slice with an index not in a list e.g.:
```python
li = [0]
print(li[2:])
```

```suggestion
            tokens = tokens[0][1:] + tokens[1:]
```

File: tests/models/llama/test_tokenization_llama.py
Comment (amyeroberts) [inline tests/models/llama/test_tokenization_llama.py:542]: What if we have two different whitespaces e.g. "   _He is not"?

File: tests/models/llama/test_tokenization_llama.py
Comment (amyeroberts) [reply tests/models/llama/test_tokenization_llama.py:569]: Still to be removed?

File: tests/models/t5/test_tokenization_t5.py
Comment (amyeroberts) [inline tests/models/t5/test_tokenization_t5.py:502]: Is it necessary to load all of this dataset for the test? It seems incredibly large

File: tests/models/t5/test_tokenization_t5.py
Comment (amyeroberts) [inline tests/models/t5/test_tokenization_t5.py:504]: Is there an explicit mention of these tests in the issue? If not, could you add it so that these aren't lost?

File: src/transformers/models/llama/tokenization_llama.py
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:95]: Could we add a `logger.warning_once` here if `legacy` is True to inform the user that legacy behaviour is being used and what that means for the outputs

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

Just to make sure I've understood - was this bug always part of the T5 tokenizer, or was it introduced in #24565? 
I'm asking because of the default value of `legacy` - at the moment any new tokenizers created will default to the legacy behaviour. If the bug was recently introduced in #24565, then it wasn't part of the most recent release and so we can default to the "new" behaviour.

File: src/transformers/utils/__init__.py
Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: Because it's messing up the import. By deleting it here, we can at least run the test individually

File: src/transformers/models/llama/tokenization_llama.py
Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py:75]: Indeed!

File: tests/models/t5/test_tokenization_t5.py
Comment (ArthurZucker) [reply tests/models/t5/test_tokenization_t5.py:504]: There is 😉

**Comment by ArthurZucker** [comment]
The bug has always been in T5, but since some models were trained with the bugged T5, we will let the user decide whether or not they incorparate the change

File: src/transformers/models/llama/tokenization_llama.py
Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py]: Actually this does not work 😅 
The reason is that when the token's length is exactly one, we get a string. Even if we wrap it around `[tokens[0][1:]] + tokens[1:]`, then there is a problem when the string is null length, it will be `[[]]`. So reverted this and left what was here! I'm pretty sure I tried both and ended up having to add the complexity!

File: tests/models/llama/test_tokenization_llama.py
Comment (amyeroberts) [reply tests/models/llama/test_tokenization_llama.py:542]: Can we add a test to make sure it's treated that way?

File: tests/models/t5/test_tokenization_t5.py
Comment (amyeroberts) [reply tests/models/t5/test_tokenization_t5.py:506]: OK, fair enough :)

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [reply tests/models/llama/test_tokenization_llama.py:542]: There is one: 
```python 
        input_ids = self.tokenizer.encode("▁He is not             ▁He")
        self.assertEqual(input_ids, [156, 46, 44, 156, 2])
        tokens = self.tokenizer.tokenize("▁He is not              ▁He")
```
This is making sure that `         ▁` is just encoded as a single ` `.

File: src/transformers/models/t5/tokenization_t5.py
Comment (amyeroberts) [inline src/transformers/models/t5/tokenization_t5.py:168]:
```suggestion
                f"You are using the legacy behaviour of the {self.__class__}. This means that tokens that come after special tokens will not be properly handled. We recommend you to"
```

File: src/transformers/models/t5/tokenization_t5.py
Comment (amyeroberts) [inline src/transformers/models/t5/tokenization_t5.py:110]: ❤️

File: src/transformers/models/llama/tokenization_llama.py
Comment (amyeroberts) [inline src/transformers/models/llama/tokenization_llama.py:120]:
```suggestion
                f"You are using the legacy behaviour of the {self.__class__}. This means that tokens that come after special tokens will not be properly handled. We recommend you to"
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this and for all the work making it safe and backwards compatible!

File: src/transformers/utils/__init__.py
Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: Having a look

Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: Also  `from transformers.utils import sentencepiece_model_pb2` works for me

Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: Comes from #24690

Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: This does not fix the version error, but fixes the issue with 3.20.3, when we cannot use seqio or anything importing protobuf: 
```python 

TypeError: Couldn't build proto file into descriptor pool!
Invalid proto descriptor for file "sentencepiece_model.proto":
  sentencepiece_model.proto: A file with this name is already in the pool.
```

Comment (ydshieh) [reply src/transformers/utils/__init__.py:180]: Hi, thanks for the information. Could you provide the full trace 🙏 . It might be useful.

Comment (ArthurZucker) [reply src/transformers/utils/__init__.py:180]: Just added it!

**Pull request merged**
